c++ SCript for the Quest Zero Tolerance and spellfix spell_zuldrak_gymers_grab

Thor1423 opened the pull request:
based on this Commit https://github.com/TrinityCore/TrinityCore/issues/30286#issue-2556389456 

I personally think that the whole thing can be designed better with c++, so here is my c++ idea

Here's my link with the C++ code and two videos
https://github.com/TrinityCore/TrinityCore/issues/30286#issuecomment-3155914134

```
UPDATE `creature_template` SET `AIName` = '', `ScriptName` = 'npc_servant_of_drakuru' WHERE `entry` = 28802;

 -- Servant of Drakuru smart ai
SET @ENTRY := 28802;
DELETE FROM `smart_scripts` WHERE `source_type` = 0 AND `entryOrGuid` = @ENTRY;

-- Servant of Drakuru Text
DELETE FROM `creature_text` WHERE `CreatureID`=@ENTRY;
INSERT INTO `creature_text` (`CreatureID`, `GroupID`, `ID`, `Text`, `Type`, `Language`, `Probability`, `Emote`, `Duration`, `Sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES
(@ENTRY, 0, 0, "Darmuk must die, mon!", 12, 0, 100, 0, 0, 0, 28884, 0, 'Servant of Drakuru - Say on Charm');
```

This is my first PR so please forgive me if I forgot something

Thor1423 committed:
Update zone_zuldrak.cpp

based on this Commit https://github.com/TrinityCore/TrinityCore/issues/30286#issue-2556389456 

I personally think that the whole thing can be designed better with c++, so here is my c++ idea

Here's my link with the C++ code and two videos
https://github.com/TrinityCore/TrinityCore/issues/30286#issuecomment-3155914134

Thor1423 committed:
Update zone_zuldrak.cpp

Fix Register AI Class

Nyr97 commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1196,6 +1196,119 @@ class spell_zuldrak_gymers_throw : public SpellScript
+
+        void SpellHit(WorldObject* caster, SpellInfo const* spell) override
+        {
+            if (!spell)
```
SpellInfo cannot ever be null here

Nyr97 commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1196,6 +1196,119 @@ class spell_zuldrak_gymers_throw : public SpellScript
+            {
+                if (caster && caster->GetTypeId() == TYPEID_PLAYER)
+                {
+                    Player* player = caster->ToPlayer();
```
this is redundant, cast ToPlayer and then check for nullability in that case

Thor1423 commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1196,6 +1196,119 @@ class spell_zuldrak_gymers_throw : public SpellScript
+
+        void UpdateAI(uint32 diff) override
+        {
+            if (_controlled)
```
Essentially, the script does exactly what it's supposed to, and therefore there were no compilation errors in the logs or in the game. 

I assumed it was solid code that I had created with the help of a friend from back then. You can see in the video that it does what it's supposed to. 

But okay, I'm not a professional programmer and I'm not afraid of corrections during the learning phase. So I'm sorry my first attempt isn't good enough for you.

Thor1423 replied:
> Just because a piece of code doesn't throw errors, exceptions, or cause crashes doesn't mean it's optimized to its best. You don't need to check that particular case every tick of Update, it's okay to delay it by 500-1000ms to optimize resources; it's not a deal of not being good enough for me, it's just a suggestion you can take or leave.

That wasn't what I meant. I'll take it to heart and try to optimize it. As I said, I'm still learning and don't know everything yet, and this is my first attempt at a PR. When I get home, I'll try to implement it, so there's a one-time delay. You're right that once the action is completed, no further updates are necessary.

Thor1423 replied:
@Nyr97 

I have now revised it and would like to submit it as is.


```c++

#include "ScriptMgr.h"
#include "Creature.h"
#include "SpellInfo.h"
#include "Unit.h"
#include "Player.h"
#include "ScriptedCreature.h"

enum
{
    SPELL_TRIGGER_ID = 52390,
    NPC_ENTRY_CONTROLLED = 28805,
    NPC_ENTRY_NORMAL = 28802,
    QUEST_ID = 12686,
    SPELL_CAST = 50361,

    EVENT_CAST_SPELL = 1,
    EVENT_CONTROL_CHECK = 2,
};

class npc_servant_of_drakuru : public CreatureScript
{
public:
    npc_servant_of_drakuru() : CreatureScript("npc_servant_of_drakuru") {}

    struct npc_servant_of_drakuruAI : public ScriptedAI
    {
        npc_servant_of_drakuruAI(Creature* creature): ScriptedAI(creature), _controller(nullptr), _controlled(false) { }

        void JustEngagedWith(Unit* /*who*/) override
        {
            _events.Reset();
            _events.ScheduleEvent(EVENT_CAST_SPELL, 4s);
            _events.ScheduleEvent(EVENT_CONTROL_CHECK, 1s);
        }

        void SpellHit(WorldObject* caster, SpellInfo const* spell) override
        {
            if (spell->Id == SPELL_TRIGGER_ID)
            {
                if (Player* player = caster->ToPlayer())
                {
                    if (player->GetQuestStatus(QUEST_ID) == QUEST_STATUS_INCOMPLETE)
                    {
                        _controller = player;
                        Talk(0);
                        me->UpdateEntry(NPC_ENTRY_CONTROLLED);
                        me->SetHealth(me->GetMaxHealth());
                        _controlled = true;
                    }
                }
            }
        }

        void UpdateAI(uint32 diff) override
        {
            _events.Update(diff);

            while (uint32 eventId = _events.ExecuteEvent())
            {
                switch (eventId)
                {
                case EVENT_CAST_SPELL:
                    if (!_controlled)
                    {
                        if (Unit* target = me->GetVictim())
                            me->CastSpell(target, SPELL_CAST, true);
                    }
                    _events.ScheduleEvent(EVENT_CAST_SPELL, 12s, 14s);
                    break;

                case EVENT_CONTROL_CHECK:
                    if (_controlled)
                    {
                        if (me->GetCharmerOrOwner() != _controller)
                        {
                            if (_controller && _controller->GetQuestStatus(QUEST_ID) == QUEST_STATUS_INCOMPLETE)
                                me->DespawnOrUnsummon(1s, 3s);

                            _controlled = false;
                            _controller = nullptr;
                            _events.Reset();
                            return;
                        }
                    }
                    _events.ScheduleEvent(EVENT_CONTROL_CHECK, 1s);
                    break;
                }
            }

            if (!_controlled)
                ScriptedAI::UpdateAI(diff);
        }

    private:
        EventMap _events;
        Player* _controller;
        bool _controlled;
    };

    CreatureAI* GetAI(Creature* creature) const override
    {
        return new npc_servant_of_drakuruAI(creature);
    }
};

void AddSC_npc_servant_of_drakuru()
{
    new npc_servant_of_drakuru();
}

```

Thor1423 committed:
Update zone_zuldrak.cpp

Code optimization and I hope it's ok this time ^^

CraftedRO commented:
you may want to put sql code in a file inside 3.3.5 update path too.

Thor1423 commented:
> you may want to put sql code in a file inside 3.3.5 update path too.

Do I have to send it separately as a PR?

Thor1423 committed:
Add files via upload

SQL for the PR Quest Zero Tolerance: https://github.com/TrinityCore/TrinityCore/pull/31196

Thor1423 commented:
ok i have added the SQL File

Thor1423 committed:
Fix Spellscript spell_zuldrak_gymers_grab

Missing Healing Effect after grabbed the Storm Cloud with this fix  now works

Thor1423 committed:
Merge pull request #1 from Thor1423/Thor1423-patch-1

Fix Spellscript spell_zuldrak_gymers_grab

Nyr97 commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1161,6 +1163,9 @@ class spell_zuldrak_gymers_grab : public SpellScript
         GetHitCreature()->CastSpell(GetCaster(), SPELL_RIDE_GYMER, args);
         GetHitCreature()->CastSpell(GetHitCreature(), SPELL_GRABBED, true);
+
+        if (GetHitCreature()->GetEntry() == NPC_STORM_CLOUD) // Storm Cloud
+        GetHitCreature()->CastSpell(GetCaster(), SPELL_HEALING_WINDS, true); // Healing Winds
```
yeah it is

Thor1423 commented:
Apparently, reading the comments here, they don't want my fixes.

It was worth a try.

Nyr97 commented:
Applying feedback that has been offered doesn't mean that it'll get reviewed the moment it happens. Merging takes time.

offl commented:
>I've made the improvement but it still doesn't seem to be good enough, so I assume there's no interest.

The fact you're having **reviews** doesn't mean someone don't want your PR to be merged, it means there are improvements to be made so it can be merged (**that's the whole purpose of making PRs**)

offl commented on sql/updates/world/3.3.5/2025_99_99_99_world.sql:
```diff
@@ -0,0 +1,10 @@
+-- Servant of Drakuru Text
+DELETE FROM `creature_text` WHERE `CreatureID`=@ENTRY;
+INSERT INTO `creature_text` (`CreatureID`, `GroupID`, `ID`, `Text`, `Type`, `Language`, `Probability`, `Emote`, `Duration`, `Sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES
+(@ENTRY, 0, 0, "Darmuk must die, mon!", 12, 0, 100, 0, 0, 0, 28884, 0, 'Servant of Drakuru - Say on Charm');
```
Missing newline

offl commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1196,6 +1201,105 @@ class spell_zuldrak_gymers_throw : public SpellScript
+
+enum
+{
+    SPELL_TRIGGER_ID = 52390,
```
Spells should have original name defined, in this case SPELL_CHARM_DRAKURU_SERVANT

offl commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1196,6 +1201,105 @@ class spell_zuldrak_gymers_throw : public SpellScript
+{
+    SPELL_TRIGGER_ID = 52390,
+    NPC_ENTRY_CONTROLLED = 28805,
+    NPC_ENTRY_NORMAL = 28802,
```
Same with NPC entries, original names

offl commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1196,6 +1201,105 @@ class spell_zuldrak_gymers_throw : public SpellScript
+    QUEST_ID = 12686,
+    SPELL_CAST = 50361,
+
+    EVENT_CAST_SPELL = 1,
```
Events should be named after spells used in events

offl commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1196,6 +1201,105 @@ class spell_zuldrak_gymers_throw : public SpellScript
+                    if (player->GetQuestStatus(QUEST_ID) == QUEST_STATUS_INCOMPLETE)
+                    {
+                        _controller = player;
+                        Talk(0);
```
Magic number, it should be defined in enum

offl commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1196,6 +1201,105 @@ class spell_zuldrak_gymers_throw : public SpellScript
+                    if (!_controlled)
+                    {
+                        if (Unit* target = me->GetVictim())
+                            me->CastSpell(target, SPELL_CAST, true);
```
Casting combat spells with triggered cast flag (true means triggered full mask) is a big no without a real reason. In this case creature will cast that spell even if polymorphed

offl commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1196,6 +1201,105 @@ class spell_zuldrak_gymers_throw : public SpellScript
+                        if (me->GetCharmerOrOwner() != _controller)
+                        {
+                            if (_controller && _controller->GetQuestStatus(QUEST_ID) == QUEST_STATUS_INCOMPLETE)
+                                me->DespawnOrUnsummon(1s, 3s);
```
You're using forceRespawnTimer but creature can't respawn (and shouldn't)
It should be just 
```cpp
me->DespawnOrUnsummon(1s);
```

offl commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1196,6 +1201,105 @@ class spell_zuldrak_gymers_throw : public SpellScript
+                            return;
+                        }
+                    }
+                    _events.ScheduleEvent(EVENT_CONTROL_CHECK, 1s);
```
_events.Repeat(1s);

offl commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1224,4 +1328,5 @@ void AddSC_zuldrak()
     RegisterSpellScript(spell_zuldrak_zuldrak_rat);
     RegisterSpellScript(spell_zuldrak_gymers_grab);
     RegisterSpellScript(spell_zuldrak_gymers_throw);
+    new npc_servant_of_drakuru();
```
Please use new register model with RegisterCreatureAI, you can find examples in this file

offl commented on sql/updates/world/3.3.5/2025_99_99_99_world.sql:
```diff
@@ -0,0 +1,10 @@
+UPDATE `creature_template` SET `AIName` = '', `ScriptName` = 'npc_servant_of_drakuru' WHERE `entry` = 28802;
+
```
SQL files should start with
```sql
--
```

offl commented on src/server/scripts/Northrend/zone_zuldrak.cpp:
```diff
@@ -1196,6 +1201,105 @@ class spell_zuldrak_gymers_throw : public SpellScript
     }
 };
 
+//Quest 12686 Zero Tolerance
```
```cpp
/*######
## Quest 12686: Zero Tolerance
######*/
```

CraftedRO replied:
this one is not a standard and can be ignored most likely.

offl replied:
Well, I use it every time
Now majority of new or updated zone scripts has it

Thor1423 committed:
Merge branch 'TrinityCore:3.3.5' into 3.3.5

Thor1423 committed:
SQL File for the Quest Dimensius the All-Devouring

SQL File for the Quest Dimensius the All-Devouring

Thor1423 committed:
Update 2025_99_99_99_2_world.sql

DELETET

Thor1423 commented:
> This PR contains code copied from [azerothcore/azerothcore-wotlk#22302](https://github.com/azerothcore/azerothcore-wotlk/pull/22302) without giving any credit to the original author or project

I can copy as much as I want
